### PR TITLE
Fix JacobianTest and CustomJacobianTest (Resolves #3470)

### LINKED ISSUE #3470: Incorrect usage of Backward method in JacobianTest
#### Issue description

The _Backward_ method(of a non-loss function layer) expects output of the _Forward_ method as the first argument.

The _JacobianTest_ and _CustomJacobianTest_ incorrectly uses the input as the first argument in [Line 77](https://github.com/mlpack/mlpack/blob/master/src/mlpack/tests/ann/ann_test_tools.hpp#L77) and [Line 128](https://github.com/mlpack/mlpack/blob/master/src/mlpack/tests/ann/ann_test_tools.hpp#L128) of [ann_test_tools.hpp](https://github.com/mlpack/mlpack/blob/master/src/mlpack/tests/ann/ann_test_tools.hpp)

**Reason**

The output of the Forward is passed back as input to the Backward in MultiLayer implementation. See Line 162 and Line 187 of [multi_layer_impl.hpp](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/multi_layer_impl.hpp#L162)

  ```
 161   for (size_t i = start + 1; i < end; ++i)
 162    network[i]->Forward(layerOutputs[i - 1], layerOutputs[i]);
 163   network[end]->Forward(layerOutputs[end - 1], output);
 ...
 ...
 186   for (size_t i = network.size() - 2; i > 0; --i)
 187    network[i]->Backward(layerOutputs[i], layerDeltas[i + 1], layerDeltas[i]);

```

**Code to verify Jacobian:** I have replicated the JacobianTest in a standalone file. 
```
#include <mlpack.hpp>

using namespace mlpack;
using namespace arma;

int main()
{
  mat input(5,1, fill::randu);
  mat output, outputHP, outputHM;
  output.set_size(5,1);
  outputHP.set_size(5,1);
  outputHM.set_size(5,1);
  
  Sigmoid sig;
  sig.Forward(input, output);

  mat derivf(5,1);
  mat oneM(5,1);
  sig.Backward(output, oneM, derivf);

  mat jacobian(input.n_elem, output.n_elem);
  mat jacobianApprox(input.n_elem, output.n_elem);
  
  double perturbation = 1e-6;

  for(size_t i = 0; i < input.n_elem; ++i)
  {
    double original = input(i);
    input(i) = original - perturbation;
    sig.Forward(input, outputHM);
    input(i) = original + perturbation;
    sig.Forward(input, outputHP);
    input(i) = original;

    outputHP -= outputHM;
    outputHP /= 2 * perturbation;
    jacobianApprox.row(i) = outputHP.t();
  }

  mat derivg(output.n_rows, output.n_cols);

  for(size_t i = 0; i < derivg.n_elem; ++i)
  {
    derivg.zeros();
    derivg(i) = 1;

    mat derivTemp;
    sig.Backward(input, derivg, derivTemp);

    // derivTemp %= derivg;

    jacobian.col(i) = derivTemp;
  }
  jacobian.print("Jacobian calculated from Backward:");
  jacobianApprox.print("Jacobian calculated using central difference of Forward:");
  mat error = abs(jacobian - jacobianApprox);
  error.print("Error:");
  std::cout << "Max Error:" <<  arma::max(arma::max(error)) << std::endl;
}
```

I ran it twice, once with Backward taking the input, and the second time using Backward taking the output for Sigmoid Layer(which is implemented using LogisticFunction). 

The output when using the _input_(which is current implementation in mlpack) is:
```
Jacobian calculated from Backward:
   0.1677        0        0        0        0
        0   0.1877        0        0        0
        0        0   0.2056        0        0
        0        0        0   0.0505        0
        0        0        0        0   0.0189
Jacobian calculated using central difference of Forward:
   0.2150        0        0        0        0
        0   0.2461        0        0        0
        0        0   0.2209        0        0
        0        0        0   0.2014        0
        0        0        0        0   0.2500
Error:
   0.0472        0        0        0        0
        0   0.0584        0        0        0
        0        0   0.0153        0        0
        0        0        0   0.1509        0
        0        0        0        0   0.2311
Max Error:0.231077
```
The Jacobians are significantly different.

The output of the above program when using the _output_ in the _Backward_:

```
From Backward:
   0.2150        0        0        0        0
        0   0.2461        0        0        0
        0        0   0.2209        0        0
        0        0        0   0.2014        0
        0        0        0        0   0.2500
Using central difference of Forward:
   0.2150        0        0        0        0
        0   0.2461        0        0        0
        0        0   0.2209        0        0
        0        0        0   0.2014        0
        0        0        0        0   0.2500
Error:
   5.8698e-11            0            0            0            0
            0   2.1528e-12            0            0            0
            0            0   4.0425e-11            0            0
            0            0            0   6.8995e-12            0
            0            0            0            0   1.9245e-11
Max Error:5.86976e-11
 
```

The error between Jacobians is in the order of 1e-11.

### PR DESCRIPTION
**Pull request by mrdaybird**
Use output of the Forward method instead of input. Fixes #3470 (See issue for more details).

### CODE REVIEW

**Review by rcurtin — COMMENTED** [review]
I think this is correct. At least, it passes all the tests, but, all layers tested with the Jacobian test do not use the first argument to `Backward()`.  So what is needed is an audit: just need to look at all the layers, and see if they all use the convention that the first parameter to `Backward()` is the *output* from `Forward()`, not the input to `Forward()`.

I'm not yet sure if this makes the Jacobian test succeed for the updated `LogSoftMax` implementation.

**Review by rcurtin — APPROVED** [review]
Went through and audited everything visually (quicker than I thought!).

 * With this change the Jacobian test for the updated `LogSoftMax` layer passes.
 * Many layers simply ignore the first parameter to `Backward()`.
 * Several layers use the first parameter just to get the batch size (`.n_cols`).
 * A couple layers (`HardTanH`, `LeakyReLU`, `PReLU`) are such that `Baclward()` is correct regardless of whether the output from or input to `Foward()` is passed.
 * I'm not 100% sure of `Softmax` but it looks like it takes the output.  A test will tell us, I guess.
 * The same is true of `Softmin`, but just glancing at the backward expression, I think it might be wrong. We'll find out when we apply a test, I guess.

I would definitely like to get these tests applied to all layers.  I was surprised to see when I went through the tests that only `Linear3D` and `LinearNoBias` are tested with the Jacobian test in the `tests/ann/layer/` directory

It could be useful in a follow-up PR to either clearly document what the first parameter to `Backward()` is, or change its name throughout the codebase, or both.  I'm not sure what the best way forward there is.

**Comment by mrdaybird** [comment]
Also to be noted, many activation functions incorrectly implement the `Deriv(x, y)` method, having the same mistake as the Backward method, consequently corresponding layers also have the incorrect `Backward` method.

I think I will make a list of all layers/functions which maybe have been incorrectly implemented, so that they can be fixed and additional tests(including the Jacobian test) be added along with the fix.

**Comment by rcurtin** [comment]
Thanks again for finding this!
